packages/forms/src/hooks/UseKeyboardNavigation.ts

```diff
@@ -0,0 +1,82 @@
+import {
+  KeyboardEvent,
+  KeyboardEventHandler,
+  RefObject,
+  useCallback,
+  useState
+} from "react";
+import { MoveDirection } from "../components/ui/text-input/SimpleTextInput";
+
+export const useKeyboardNavigation = (
+  ref: RefObject<HTMLInputElement>,
+  /**
+   * User-provided onKeyDown. Internal handler should forward calls to this.
+   * */
+  onKeyDown?: KeyboardEventHandler<HTMLInputElement>,
+  onEnter?: () => void,
+  onEsc?: () => void,
+  /**
+   * onMove callback, triggered when user tries to move outside of field using arrow keys, tab or shift+tab.
+   * */
+  onMove?: (direction: MoveDirection) => void
+) => {
+  const [wasCancelled, setWasCancelled] = useState(false);
+
+  const blurMoveAndCancel = useCallback(
+    (direction: MoveDirection, e: KeyboardEvent<HTMLInputElement>) => {
+      ref.current!.blur();
+      if (onMove) {
+        onMove(direction);
+      }
+      e.preventDefault();
+      e.stopPropagation();
+    },
+    [onMove, ref]
+  );
+
+  const onKeyDownHandler: KeyboardEventHandler<HTMLInputElement> = useCallback(
+    ev => {
+      const { key } = ev;
+      if (key === "Enter") {
+        ref.current!.blur();
+        if (onEnter) {
+          onEnter();
+        }
+      } else if (onEsc && key === "Escape") {
+        setWasCancelled(true);
+        setTimeout(() => {
+          onEsc(); // Do this after set state is done. Is there a better way?
```

Comment:
`useEffect` with `wasCancelled` as dependency?

Reply:
Hm, that might work? Should try! 
I'd rather do it in a separate branch though, this is just a refactoring of already existing code.